src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Debugging/RazorBreakpointResolver.cs

```diff
@@ -20,20 +25,80 @@ namespace Microsoft.VisualStudio.Razor.LanguageClient.Debugging;
 internal class RazorBreakpointResolver(
     FileUriProvider fileUriProvider,
     LSPDocumentManager documentManager,
-    ILSPBreakpointSpanProvider breakpointSpanProvider) : IRazorBreakpointResolver
+    ILSPBreakpointSpanProvider breakpointSpanProvider,
+    LanguageServerFeatureOptions languageServerFeatureOptions,
+    IRemoteServiceInvoker remoteServiceInvoker) : IRazorBreakpointResolver
 {
-    private record CacheKey(Uri DocumentUri, long? HostDocumentSyncVersion, int Line, int Character);
+    private record CohostCacheKey(DocumentId DocumentId, VersionStamp Version, int Line, int Character) : CacheKey;
+    private record LspCacheKey(Uri DocumentUri, long? HostDocumentSyncVersion, int Line, int Character) : CacheKey;
+    private record CacheKey;
 
     private readonly FileUriProvider _fileUriProvider = fileUriProvider;
     private readonly LSPDocumentManager _documentManager = documentManager;
     private readonly ILSPBreakpointSpanProvider _breakpointSpanProvider = breakpointSpanProvider;
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions;
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
 
     // 4 is a magic number that was determined based on the functionality of VisualStudio. Currently when you set or edit a breakpoint
     // we get called with two different locations for the same breakpoint. Because of this 2 time call our size must be at least 2,
     // we grow it to 4 just to be safe for lesser known scenarios.
     private readonly MemoryCache<CacheKey, Range> _cache = new(sizeLimit: 4);
 
-    public async Task<Range?> TryResolveBreakpointRangeAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+    public Task<Range?> TryResolveBreakpointRangeAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+    {
+        if (_languageServerFeatureOptions.UseRazorCohostServer)
+        {
+            return TryResolveBreakpointRangeViaCohostingAsync(textBuffer, lineIndex, characterIndex, cancellationToken);
+        }
+
+        return TryResolveBreakpointRangeViaLspAsync(textBuffer, lineIndex, characterIndex, cancellationToken);
```

Comment:
Nit: Ternary conditional operator could simplify this, unless you strongly dislike it.